Library/ViewModels/DiscoveryPageViewModel.swift

```diff
@@ -142,36 +145,58 @@ public final class DiscoveryPageViewModel: DiscoveryPageViewModelType, Discovery
       .map { $0.1 }
 
     let requestFirstPageWith = Signal.merge(
-        firstPageParams,
-        (firstPageParams.takeWhen(environmentChanged)))
+      firstPageParams,
+      (firstPageParams.takeWhen(environmentChanged)),
+      firstPageParams.takeWhen(self.pulledToRefreshProperty.signal))
 
     let paginatedProjects: Signal<[Project], NoError>
     let pageCount: Signal<Int, NoError>
-    (paginatedProjects, self.projectsAreLoading, pageCount) = paginate(
+    let isLoading: Signal<Bool, NoError>
+    (paginatedProjects, isLoading, pageCount) = paginate(
       requestFirstPageWith: requestFirstPageWith,
       requestNextPageWhen: isCloseToBottom,
-      clearOnNewRequest: true,
+      clearOnNewRequest: false,
       skipRepeats: false,
       valuesFromEnvelope: { $0.projects },
       cursorFromEnvelope: { $0.urls.api.moreProjects },
       requestFromParams: { AppEnvironment.current.apiService.fetchDiscovery(params: $0) },
       requestFromCursor: { AppEnvironment.current.apiService.fetchDiscovery(paginationUrl: $0) },
       concater: { ($0 + $1).distincts() })
 
-      let projects = Signal.merge(
-        paginatedProjects,
-        self.selectedFilterProperty.signal.skipNil().skipRepeats().mapConst([])
+    let projects = Signal.merge(
+      paginatedProjects,
+      self.selectedFilterProperty.signal.skipNil().skipRepeats().mapConst([])
       )
       .skip { $0.isEmpty }
       .skipRepeats(==)
 
-      self.projectsLoaded =  self.selectedFilterProperty.signal
-        .takePairWhen(projects)
-        .map { ($1, $0) }
+    self.projectsLoaded =  self.selectedFilterProperty.signal
+      .takePairWhen(projects)
+      .map { ($1, $0) }
 
     self.asyncReloadData = self.projectsLoaded.take(first: 1).ignoreValues()
 
-    self.showEmptyState = Signal.combineLatest(paramsChanged, self.projectsAreLoading, paginatedProjects)
+    let isRefreshing = isLoading
+      .combineLatest(with: self.pulledToRefreshProperty.signal)
+      .map { $0.0 }
+      .skipRepeats()
+
+    let projectsLoadingNoRefresh = Signal.merge(
+      isLoading,
+      isLoading.takeWhen(isRefreshing).mapConst(false)
```

Comment:
Isn't this `isLoading` duplicate signal? Can you explain this?

Reply:
While `isLoading` will get updated every time the projects are loading or stop loading, the `isLoading.takeWhen(isRefreshing).mapConst(false)` signal will only emit once the user pulls to refresh. Without the `isLoading`, the `projectsLoadingNoRefresh` won't emit until the user pulls to refresh and the UIRefreshControl won't spin. Does it make sense?